action create-hardfork cmd + update description fields in all governance action files.

CarlosLopezDeLara opened the pull request:
# Changelog

```yaml
- description: |
    Introduces the `governance action create-hardfork` cmd.
# uncomment types applicable to the change:
  type:
  - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - improvement    # QoL changes e.g. refactoring
  # - bugfix         # fixes a defect
  - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

Closes https://github.com/IntersectMBO/cardano-cli/issues/721
 
# How to trust this PR

Run the command 
```
cardano-cli conway governance action create-hardfork \
--mainnet \
--anchor-url example.com \
--anchor-data-hash c7ddb5b493faa4d3d2d679847740bdce0c5d358d56f9b1470ca67f5652a02745 \
--deposit-return-stake-verification-key-file stake.vkey \
--governance-action-deposit 12345 \
--protocol-major-version 10 \
--protocol-minor-version 0 \
--out-file conway-create-hardfork.action
```
Explore the text envelope en check it with cbor.me 
```
{
    "type": "Governance proposal",
    "description": "Hardfork initiation governance action",
    "cborHex": "84193039581de18f4a3466a404c11eb410313015b88e447d81b60089e25f611600e6058301f6820a00826b6578616d706c652e636f6d5820c7ddb5b493faa4d3d2d679847740bdce0c5d358d56f9b1470ca67f5652a02745"
}
```
```
[12345, h'E18F4A3466A404C11EB410313015B88E447D81B60089E25F611600E605', [1, null, [10, 0]], ["example.com", h'C7DDB5B493FAA4D3D2D679847740BDCE0C5D358D56F9B1470CA67F5652A02745']]
```

or with 
```
cardano-cli conway governance action view --action-file conway-create-hardfork.action
{
    "anchor": {
        "dataHash": "c7ddb5b493faa4d3d2d679847740bdce0c5d358d56f9b1470ca67f5652a02745",
        "url": "example.com"
    },
    "deposit": 12345,
    "governance action": {
        "contents": [
            null,
            {
                "major": 10,
                "minor": 0
            }
        ],
        "tag": "HardForkInitiation"
    },
    "return address": {
        "credential": {
            "keyHash": "8f4a3466a404c11eb410313015b88e447d81b60089e25f611600e605"
        },
        "network": "Mainnet"
    }
}
```
# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [x] Self-reviewed the diff

<!-- 
### Note on CI ###
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges. Please contact IOG node developers to do this
for you. 
-->

carbolymer reviewed (approved):
LGTM, nice job! Can you add a test for that, similarly to the other ones here: https://github.com/IntersectMBO/cardano-cli/blob/main/cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/Action.hs

smelc commented on cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Actions.hs:
```diff
@@ -385,4 +386,48 @@ pNetwork  = asum $ mconcat
+              <*> pPV
+              <*> pFileOutDirection "out-file" "Output filepath of the no confidence proposal."
+        )
+    $ Opt.progDesc "Create a hardfork inititation proposal."
```
```suggestion
    $ Opt.progDesc "Create a hardfork initiation proposal."
```

smelc commented on cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Actions.hs:
```diff
@@ -385,4 +386,48 @@ pNetwork  = asum $ mconcat
+              <*> pAnchorUrl
+              <*> pAnchorDataHash
+              <*> pPV
+              <*> pFileOutDirection "out-file" "Output filepath of the no confidence proposal."
```
```suggestion
              <*> pFileOutDirection "out-file" "Output filepath of the create hardfork proposal."
```

smelc commented on cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/Actions.hs:
```diff
@@ -376,3 +379,41 @@ runGovernanceActionTreasuryWithdrawalCmd
+      , Cmd.returnStakeAddress
+      , Cmd.mPrevGovernanceActionId
+      , Cmd.proposalUrl
+      , Cmd.proposalHash
```
```suggestion
      , Cmd.proposalHash = anchorDataHash
```

goes with the punning below

smelc commented on cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/Actions.hs:
```diff
@@ -376,3 +379,41 @@ runGovernanceActionTreasuryWithdrawalCmd
+
+  let proposalAnchor = L.Anchor
+        { L.anchorUrl = unProposalUrl proposalUrl
+        , L.anchorDataHash = proposalHash
```
```suggestion
        , L.anchorDataHash
```

smelc commented on cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/Actions.hs:
```diff
@@ -376,3 +379,41 @@ runGovernanceActionTreasuryWithdrawalCmd
+
+  firstExceptT GovernanceActionsCmdWriteFileError . newExceptT
+    $ conwayEraOnwardsConstraints eon
+    $ writeFileTextEnvelope outFile Nothing proposalProcedure
```
Could we replace the `Nothing` by something, so that the generated proposal has an helpful description text?

CarlosLopezDeLara replied:
addressed in latest commit

Would be good to do the same on all other action files.

smelc reviewed (changes requested):
LGTM :champagne: A few typos in the help needs fixing.

Jimbo4350 reviewed (commented):
LGTM! Feel free to merge after @smelc approves.

Jimbo4350 reviewed (changes requested):
The commit history needs to be cleaned up. Squashing commits and leaving a commit message like: 
```
Squashed commit of the following:

commit https://github.com/IntersectMBO/cardano-cli/commit/82cc5c1ab7ff50d4072a4e4dea0ba9f6665e04be
Merge: https://github.com/IntersectMBO/cardano-cli/commit/8229fc0c0e3d07f26f1c0707893cf5544746e115 https://github.com/IntersectMBO/cardano-cli/commit/e3dc0a02ffdcd44779a7a7dac99d68d1cc235bdc
Author: Carlos LopezDeLara <carlos.lopezdelara@iohk.io>
Date:   Thu May 2 13:10:52 2024 -0600

    Merge branch 'cl/hardforkinit' of github.com:IntersectMBO/cardano-cli into cl/hardforkinit

commit https://github.com/IntersectMBO/cardano-cli/commit/8229fc0c0e3d07f26f1c0707893cf5544746e115
Author: Carlos LopezDeLara <carlos.lopezdelara@iohk.io>
Date:   Wed May 1 12:09:49 2024 -0600

    action create-hardfork command

    Co-Authored-By: Clément Hurlin <smelc@users.noreply.github.com>

commit https://github.com/IntersectMBO/cardano-cli/commit/e3dc0a02ffdcd44779a7a7dac99d68d1cc235bdc
Author: Carlos LopezDeLara <carlos.lopezdelara@iohk.io>
Date:   Wed May 1 12:09:49 2024 -0600

    action create-hardfork command

    Co-Authored-By: Clément Hurlin <smelc@users.noreply.github.com>

@CarlosLopezDeLara
```
is not very readable. When you squash commits you need to summarize the changes in those commits in a concise and clear manner. 

I also don't know what is happening in `3d8a21f4792c0983b94f493ee064a3203a1c86da`.

CarlosLopezDeLara committed:
- Implements action create-hardfork command.

- Implements action create-hardfork command
- Update golden files
- Include a description on the "description field" of all governance action commands
- Export a friendly function that returns a ByteString (so not in IO)
- Export buildShelleyAddress

Jimbo4350 reviewed (approved):
LGTM!
